app/helpers/proxmox_vm_attrs_helper.rb

```diff
@@ -0,0 +1,107 @@
+# frozen_string_literal: true
+
+# Copyright 2018 Tristan Robert
+
+# This file is part of ForemanFogProxmox.
+
+# ForemanFogProxmox is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# ForemanFogProxmox is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with ForemanFogProxmox. If not, see <http://www.gnu.org/licenses/>.
+
+require 'fog/proxmox/helpers/disk_helper'
+require 'fog/proxmox/helpers/nic_helper'
+require 'fog/proxmox/helpers/cpu_helper'
+require 'foreman_fog_proxmox/value'
+require 'foreman_fog_proxmox/hash_collection'
+include ForemanFogProxmox::ProxmoxComputeAttributes
+# Convert a foreman form server hash into a fog-proxmox server attributes hash
+module ProxmoxVmAttrsHelper
+  def object_to_attributes_hash(vm, from_profile)
+    paramScope = from_profile ? "compute_attribute[vm_attrs]" : "host[compute_attributes]"
+    vm_h = ActiveSupport::HashWithIndifferentAccess.new
+    keys = [ :vmid, :node_id, :type, :pool]
+    extra_keys = [:cpu_type]
+    main = vm.attributes.select { |key, _value| keys.include? key }
+    interfaces = Hash[vm.config.interfaces.each_with_index.map do |interface, idx|
+	    [idx.to_s, interface_compute_attributes(interface.attributes)]
+                                               end ]
+    vm.config.all_attributes.each do |key, value|
+      if !keys.include? key
+        vm_h.merge!({key => {:name => "#{paramScope}[config_attributes][#{key}]", :value => value}})
+      end
+    end
+    main.each do |key, value|
+      vm_h.merge!({key => {:name => "#{paramScope}[#{key}]", :value => value}})
+    end
+    logger.warn("*********************8 interfaces #{vm.interfaces}")
+    logger.warn("******************* net attrs #{network_attrs(paramScope, vm.interfaces)}")
+    vm_h.merge!({ :interfaces => network_attrs(paramScope, vm.interfaces) })
+    vm_h.merge!({:pool => {:name => "#{paramScope}[pool]", :value => vm.pool}})
+    vm_h.merge!({:cpu_type => {:name => "#{paramScope}[config_attributes][cpu_type]", :value => vm.config.cpu_type}})
+    vm_h.merge!({ :disks => volumes_attrs(paramScope, vm.volumes) })
```

Comment:
Would it be more readable to use `[]` instead of `merge!()` (also for the other occurrences)?
```suggestion
    vm_h[:interfaces] = network_attrs(paramScope, vm.interfaces)
    vm_h[:pool] = {
      name: "#{paramScope}[pool]",
      value: vm.pool,
    }
    vm_h[:cpu_type] = {
      name: "#{paramScope}[config_attributes][cpu_type]",
      value: vm.config.cpu_type
    }
    vm_h[:disks] = volumes_attrs(paramScope, vm.volumes)
```
If not, we could reduce this to one bigger `merge!()` since the keys do not seem to collide.